### PR TITLE
exceptions for compute path through poses

### PR DESCRIPTION
**Pull request by jwallace42**
---

## Basic Info

| Info | Please fill out this column |
| ------ | -----  |
| Ticket(s) this addresses   | #3043  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Gazebo |

---

## Description of documentation updates required from your changes
Added error codes for compute path through poses. 

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_planner/src/planner_server.cpp [hunk 1]
```diff
@@ -336,11 +336,13 @@ void PlannerServer::computePlanThroughPoses()
 
   auto start_time = steady_clock_.now();
 
-  // Initialize the ComputePathToPose goal and result
+  // Initialize the ComputePathThroughPoses goal and result
   auto goal = action_server_poses_->get_current_goal();
   auto result = std::make_shared<ActionThroughPoses::Result>();
   nav_msgs::msg::Path concat_path;
 
+  geometry_msgs::msg::PoseStamped curr_start, curr_goal;
```
Comment (SteveMacenski) [inline nav2_planner/src/planner_server.cpp:344]: Move back to original location near where its first use is

File: nav2_planner/src/planner_server.cpp [hunk 2]
```diff
@@ -350,11 +352,12 @@ void PlannerServer::computePlanThroughPoses()
 
     getPreemptedGoalIfRequested(action_server_poses_, goal);
 
-    if (goal->goals.size() == 0) {
+    if (goal->goals.empty()) {
       RCLCPP_WARN(
         get_logger(),
-        "Compute path through poses requested a plan with no viapoint poses, returning.");
+        "Compute path through poses requested a plan with no waypoint poses, returning.");
```
Comment (SteveMacenski) [inline nav2_planner/src/planner_server.cpp:358]: Would this be what throws NoWaypointsGiven?

File: nav2_msgs/action/ComputePathThroughPoses.action [hunk 3]
```diff
@@ -1,3 +1,16 @@
...
+int16 NONE=0
+int16 UNKNOWN=1
+int16 TF_ERROR=2
+int16 START_OUTSIDE_MAP=3
+int16 GOAL_OUTSIDE_MAP=4
+int16 START_OCCUPIED=5
+int16 GOAL_OCCUPIED=6
+int16 TIMEOUT=7
+int16 NO_VALID_PATH=8
+int16 NO_WAYPOINTS_GIVEN=9
```
Comment (SteveMacenski) [inline nav2_msgs/action/ComputePathThroughPoses.action:12]: I prefer the via point language when talking about points going through states continuously. I think "waypoints" is a loaded term that people usually think of as things to stop at. Given we have the waypoint follower that stops at poses, I think having another name is helpful to separate these ideas

File: nav2_planner/src/planner_server.cpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_planner/src/planner_server.cpp:358]: See viapoint comment above too

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_planner/src/planner_server.cpp (hunk 1 above)
Comment (jwallace42) [reply nav2_planner/src/planner_server.cpp:344]: The curr_start and goal need to be outside of the try catch in order to print out the offending start/goal if an exception is thrown.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Review by SteveMacenski — APPROVED** [review]
LGTM, anything else before merge? I retriggered CI, doesn't seem like it went the last time

**Comment by jwallace42** [comment]
Nope! I think this is good to go.

**Pull request merged**